### DISCUSSION TITLE
How to store instrument settings or metadata to file

### DISCUSSION THREAD

**Discussion by bleykauf** [pymeasure/pymeasure#501] [Q&A]
I am trying to store instrument settings to a csv file within a Procedure via Results + Worker (just as in the tutorial examples), similar to how Parameters are stored. More explicitly, I want to read these parameters from an instrument when the `execute` method of the procedure is executed.

I've tried to use the Parameters in the `startup` method and overwrite them in the `execute` method but to no avail: It always stores the default value. As the documentation states, Parameters are used as inputs only. 

Is there any way to achieve this?

**Comment by msmttchr** [comment]
Could you try declaring the parameter of class Measurable? This may work,
even though a proper documentation does not currently exists. Have a look
at file pymeasure/experiments/parameters.py and class Measurable.

If it does work, you could submit a PR for documentation update.

Il Gio 21 Ott 2021, 16:51 Bastian Leykauf ***@***.***> ha
scritto:

**Reply by bleykauf** [reply]
Thanks for your help. Unfortunately, this doesn't really do what I want. Here is the code I am using (following the example code from the repo):

```python
import logging
import random
import sys
from time import sleep

from pymeasure.display.Qt import QtGui
from pymeasure.display.windows import ManagedWindow
from pymeasure.experiment import (
    FloatParameter,
    IntegerParameter,
    Parameter,
    Procedure,
    Results,
)
from pymeasure.experiment import Measurable


from pymeasure.experiment.results import unique_filename


log = logging.getLogger("")
log.addHandler(logging.NullHandler())


class TestProcedure(Procedure):

    iterations = IntegerParameter("Loop Iterations", default=10)
    delay = FloatParameter("Delay Time", units="s", default=0.1)
    seed = Parameter("Random Seed", default="12345")
    iteration = Measurable("Iteration", default=0)
    random_number = Measurable("Random Number")
    magic_number = Measurable("Magic Number")

    magic_number = Measurable("Magic Number")

    DATA_COLUMNS = ["Iteration", "Random Number"]

    def startup(self):
        log.info("Setting up random number generator")
        random.seed(self.seed)

    def measure(self):
        data = self.get_datapoint()
        data["Random Number"] = random.random()
        log.debug("Produced numbers: %s" % data["Random Number"])
        self.emit("results", data)
        self.emit("progress", 100.0 * self.iteration.value / self.iterations)

    def execute(self):

        log.info("Set the magic number")
        # In reality, we would read out some instrument settings.
        self.magic_number.value = 42

        log.info("Starting to generate numbers")
        for self.iteration.value in range(self.iterations):
            self.measure()
            sleep(self.delay)
            if self.should_stop():
                log.warning("Catch stop command in procedure")
                break

    def shutdown(self):
        log.info("Finished")


class MainWindow(ManagedWindow):
    def __init__(self):
        super(MainWindow, self).__init__(
            procedure_class=TestProcedure,
            inputs=["iterations", "delay", "seed"],
            displays=["iterations", "delay", "seed"],
            x_axis="Iteration",
            y_axis="Random Number",
            directory_input=True,
        )
        self.setWindowTitle("GUI Example")

    def queue(self):
        directory = self.directory
        filename = unique_filename(directory)

        procedure = self.make_procedure()
        results = Results(procedure, filename)
        experiment = self.new_experiment(results)

        self.manager.queue(experiment)


if __name__ == "__main__":
    app = QtGui.QApplication(sys.argv)
    window = MainWindow()
    window.show()
    sys.exit(app.exec_())

```

When I execute the measurement, I get this CSV file:

```
#Procedure: <__main__.TestProcedure>
#Parameters:
#	Delay Time: 0.1 s
#	Loop Iterations: 10
#	Random Seed: 12345
#Data:
Iteration,Random Number
0,0.45697818440486826
1,0.7308852116749263
2,0.4293074483607703
3,0.4517169213833959
4,0.11134793367233986
5,0.602626058001951
6,0.4033699392617659
7,0.11288375441323473
8,0.06577890775332551
9,0.6567558811578444
```

That is, the Magic Number does not appear in the header.

I could add `"Magic Number"` to `DATA_COLUMNS` but then I just have 42 repeated in every single line just like the random number.

**Comment by msmttchr** [comment]
I think you are right, possibly `Measurable ` is not the right method to achieve this. Indeed, if I get it right, you would like to display a sort of test condition field in the log header to display information about the test. For example, the serial number of the instrument or the time when the measurement was done or any other thing that does not need to be repeated for each data point.

If this the case, in your example, you would like something like:
```
#Procedure: <__main__.TestProcedure>
#Parameters:
#	Delay Time: 0.1 s
#	Loop Iterations: 10
#	Random Seed: 12345
#Test condition:
#	Magic Number 42

#Data:
Iteration,Random Number
0,0.45697818440486826
1,0.7308852116749263
.
.
.

```

I think this is not achievable with current feature of `PyMeasure`, but the right location to look at, it would be the `header` method of `Results` class (see https://github.com/pymeasure/pymeasure/blob/master/pymeasure/experiment/results.py). Extending this method should allow you to add further information (be careful that also the log file is read when you use the open button in the GUI and changing the format of the log file may break also this function).

Hope this may help.

**Reply by bleykauf** [reply]
Just to clarify: I would like to store instrument parameters etc. in den csv. This could be a serial number etc as you say but also instrument settings.

It would be ideal if I could decide whether I want to set some Parameters or if I want to read them from the instrument. This would be useful for example when adjusting a trace on a spectrum analyzer using the manual buttons on the device and then recording that trace with pymeasure. In this case I do want to save the Video Bandwidth of the device to the csv, without explicitly setting it for my hypothetical `SaveSpectrumAnalyzerTraceProcedure`.

**Reply by CasperSchippers** [reply]
Indeed, at this moment, this is not a built-in feature, though I think it would be a nice addition. (I would for example also like to be able to do something like this indeed, to store some settings of the instruments or surrounding conditions)
One thing to keep in mind for such an addition is to consider when it written to the file: the header is written when the experiment is queued, but I think it would be nice to write this secondary part only when the measurement actually commences (this allows you to more accurately store the settings or some other conditions like the exact time when it starts or so)

**Reply by CasperSchippers** [reply]
Now I come to think of it, this should happen after the `startup` method has ran, as it will require communication with instruments.

**Reply by CasperSchippers** [reply]
I have opened a PR to address this suggestion. @bleykauf, could you have a look and see if this suits your needs.
